mitodl/mitxpro
# API view for creating coupons

Based on #115

#### Pre-Flight checklist

#### What are the relevant tickets?
Partially addresses #107 and #15

#### What's this PR do?
- Adds an API view for creating coupons
- Adds an API view for listing products (needed by coupon creation form)
- Adds a view for downloading a csv of coupon codes associated with a `CouponPaymentVersion`

#### How should this be manually tested?
- Log in as a staff user.
- Post data like this to `/api/coupons` for creating multiple single-use coupons:
```
{                       "tag": "FAKETAG", 
                         "name": "FAKENAME",
                         "automatic": true,
                         "activation_date": "2018-01-01T00:00:00Z",
                         "expiration_date": "2019-12-31T00:00:00Z",
                         "amount": 0.75,
                         "num_coupon_codes": 2,
                         "coupon_type": "single-use",
                         "company": "Acme Corp.",
                         "payment_type": "credit_card",
                         "payment_transaction": "fakeorderid",
                         "products": [1]
 }
```
This should generate 2 `Coupon` objects with different random `coupon_code`s, 2 `CouponVersion`s, 1 `CouponPayment`, 1 `CouponPaymentVersion`, 1 `CouponPayment`, and 2 `CouponEligibility` objects.  


- Post something like this for a promo coupon:
```
{                       "tag": "newcoupon2", 
                         "name": "newname2",
                         "coupon_code": "NEWCODE2",
                         "automatic": false,
                         "activation_date": "2018-01-01T00:00:00Z",
                         "expiration_date": "2019-12-31T00:00:00Z",
                         "amount": 0.75,
                         "coupon_type": "promo",
                         "max_redemptions": 1000,
                         "company": "Acme Corp.",
                         "payment_type": "credit_card",
                         "payment_transaction": "123asd456",
                          "products": [1]
 }
```
This should generate 1 `Coupon` objects with `coupon_code` of 'NEWCODE2', 1 `CouponVersion`, 1 `CouponPayment`, 1 `CouponPaymentVersion`, 1 `CouponPayment`, and 1 `CouponEligibility` object.

- Try the URL `/couponpayment/codes/<couponpaymentversion_id>` using one of the `CouponPaymenVersion` id's generated above.  A URL containing the coupon code(s) should be downloaded.

- None of the above should work if you are not logged in as a staff/admin user.

- Go to `/api/products`, you should get a JSON list of products that includes `product_type` and `title`
